# 단일 표본 t 검정
- 표본 데이터 로딩
- 표본 데이터가 정규성 검정하기 
- 단일표본 t검정
    - p-value < 0.05 : 귀무가설 기각 대립가설 채택 통계적으로 유의미하다고 해석함
    - p-value >= 0.05 : 귀무가설 채택 대립가설 기각 

- txt 파일의 데이터를 읽어서 float형으로 형변환
- 정규성 검정
    - kstest를 활용해 정규성 검정 진행


In [9]:
# data 불러오기
with open("./datas2/성인여성_키_데이터.txt", "r") as f:
    data = f.read().split('\n')
    # print(data)
    # map(data_type, list_data)
    # print(map(float,data)) # <map object at 0x74a9143bee30>
    data = list(map(float, data)) # list로 형변환
    print(data)
    # print(len(data))

FileNotFoundError: [Errno 2] No such file or directory: './datas2/성인여성_키_데이터.txt'

In [ ]:
data[:3]

[150.27, 142.94, 160.99]

In [ ]:
from scipy import stats
import numpy as np

# 평균과 표준편차 계산
mean_value = np.mean(data)
std_dev = np.std(data, ddof=0)  # 모집단 표준편차(기본 설정)

mean_value, std_dev

(np.float64(156.9332), np.float64(9.974187774450611))

## 가설 설정
[문제]

귀무가설: 표본 데이터의 평균은 163과 같다.

대립가설: 표본 데이터의 평균은 163과 다르다.

## 정규성 검정 
- 수집한 표본 데이터가 정규 분포를 따르는지 검증한다.
- 방법 :KS-test
### 정규성 검정의 가설
- 귀무가설 : 수집한 표본의 데이터는 정규성을 따른다.
- 대립가설 : 수집한 표본의 데이터는 정규성을 따르지않는다.

In [ ]:
result = stats.kstest(data,stats.norm.cdf, args=(mean_value,std_dev))

print(result)

KstestResult(statistic=np.float64(0.10925974986314058), pvalue=np.float64(0.8950958424230889), statistic_location=np.float64(153.07), statistic_sign=np.int8(-1))


[결과 해설]
- p-value > 0.05: 귀무가설을 기각하지 못함 → 데이터가 정규분포를 따르지 않는다 라고 말할 충분한 증거가 없음.
- p-value <= 0.05: 귀무가설을 기각 → 데이터는 정규분포를 따르지 않는다고 판단.

- 방법2 : Shapiro-Wilk 검정
    - scipy.stats.shapiro()

In [ ]:
stats.shapiro(data) 

ShapiroResult(statistic=np.float64(0.9535807148228921), pvalue=np.float64(0.3014257453674708))

[해설]

p-value > 0.05 : 정구분포를 따르지 않는다라고 말할 충분한 증거가 없음. 귀무가설 기각하지 않음


[문제]

귀무가설: 표본 데이터의 평균은 163과 같다.

대립가설: 표본 데이터의 평균은 163과 다르다.

In [ ]:
# 단일 표본 t 검정 수행
# 검중 수치 = 163, 귀무가설 평균키는 163이다.

print(stats.ttest_1samp(data, 163)) 

TtestResult(statistic=np.float64(-2.979804412662668), pvalue=np.float64(0.00651044533584796), df=np.int64(24))


### 결과 해석
- pvalue=np.float64(0.00651044533584796) p_value < 0.05 즉, 기무가설 기각한다. 

대립가설 채택한다. 

- 기무가설 : 표본 데이터의 평균은 163과 같다. 기각

- 기무가설, 대립가설 다시 다시 세워서 해야함

- 결과적으로 여성 키의 표본 데이터의 평균은 163과 다르다. 라는 것이 통계적으로 유의미 하다. 

# 독립 표본 t- 검정

- 두 그룹 A, B반의 점수 차이가 있는지 판단하기

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df1 = pd.read_csv("./datas2/반별_점수_type1.csv", encoding = "euc-kr")
# df1 = pd.read_csv("./datas2/반별_점수_type1.csv")
df1.head()

# euc-kr 

,반,점수
0,A,73
1,A,69
2,A,71
3,A,71
4,A,73


In [ ]:
# A반의 데이터만 추출
group_A = df1['점수'].loc[df1['반'] == 'A'].values
# loc : 위치값으로 필터링 할 때 사용

# B반의 데이터만 추출
group_B = df1['점수'].loc[df1['반'] == 'B'].values

print(group_A)
print(group_B)

[73 69 71 71 73 67 73 69 62 74 68 66 70 82 70 65 76 73 58 81]
[63 56 73 61 55 77 75 65 61 55]


In [ ]:
df1['점수'].loc[df1['반'] == 'B'].values
# 칼럼 이름 할때는 대괄호를 해야함 
# 칼럼 이름의 벨류값이 A와 같느냐




array([63, 56, 73, 61, 55, 77, 75, 65, 61, 55])

- 위 두 그룹의 각각의 평균과 표준편차 구하기

In [ ]:
import numpy as np

# A반의 평균과 표준 편차
mean_A = np.mean(group_A)
std_A = np.std(group_A, ddof=1)  # 표본 표준 편차 (ddof=1)

# B반의 평균과 표준 편차
mean_B = np.mean(group_B)
std_B = np.std(group_B, ddof=1)  # 표본 표준 편차 (ddof=1)

# 결과 출력
print(f"A반 평균: {mean_A:.2f}, 표준 편차: {std_A:.2f}")
print(f"B반 평균: {mean_B:.2f}, 표준 편차: {std_B:.2f}")

A반 평균: 70.55, 표준 편차: 5.68
B반 평균: 64.10, 표준 편차: 8.28


# 정규성 검정
- 정규성 검정 가설수립
    - 영가설 : 두 그룹의 표본 데이터는 정규분포를 따른다.
    - 독립가설 : 두 그룹의 표본 데이터는 정규성를 따르지 않는다


[kstest]

[ Shapiro-Wilk]

In [11]:
# kstest
from scipy import stats

# group_A와 group_B의 평균과 표준편차를 사용하여 정규성 검정 수행
print(stats.kstest(
  group_A, 
  stats.norm.cdf, 
  args=(np.mean(group_A), np.std(group_A, ddof=0))))

print(stats.kstest(
  group_B, 
  stats.norm.cdf, 
  args=(np.mean(group_B), np.std(group_B, ddof=0))))

KstestResult(statistic=np.float64(0.1290433386337495), pvalue=np.float64(0.8515822805406548), statistic_location=np.float64(73.0), statistic_sign=np.int8(1))
KstestResult(statistic=np.float64(0.17142174723652215), pvalue=np.float64(0.8842512533111235), statistic_location=np.float64(73.0), statistic_sign=np.int8(-1))


[해석]

p-value >= 0.05 : 모두 정규 분포를 따른다고 판단 , 기무가설 기각 X 통계학 적으로 유의하지 않다. 즉, 정규성을 따른다.

독립표본 t검정 할 수 있음.

방법2 : Shapiro-Wilk 검정

scipy.stats.shapiro(d)


# 등분산성 검정

- 가설 수립
    - 영가설(귀무가설) : 두 그룹은 분산이 유사함(등분산성 , 등분산을 따른다.)
    - 대립가설 : 두 그룹은 분산이 유사하지 않다.(등분산을 따르지 않는다.)



In [12]:
stats.levene(group_A,group_B)

LeveneResult(statistic=np.float64(2.033067087400979), pvalue=np.float64(0.1649640862221014))

[해석]

p-value >= 0.005 : 두 그룹은 등분산성을 따른다.

그래서 독립검정 T검정 가능

# 독립표본 t-검정 수행

- 가설 수립
    - 영가설(귀무가설) : 두 그룹의 표본 평균은 같다. 
    - 대립가설 : 두 그룹의 표본 평균은 다르다. 

In [16]:
# A와 B간에는 차이가 존재함을 확인
# equal_var = True(등분산 만족)
stats.ttest_ind(group_A,group_B, equal_var=True)

TtestResult(statistic=np.float64(2.5128526794964134), pvalue=np.float64(0.018010953528937678), df=np.float64(28.0))

[해석]

p-value < 0.05 : 그룹 A와 그룹 B의 평균 차이가 있다는 것을 알 수 있음

기무가설 기각, 대립가설 재택 


통계값이 양수이므로 : A반 학생들의 성적이 더 좋다는 것을 알 수 있음

## Tip. 다른 데이터 포맷의 처리
A반, B반 두개의 컬럼으로 데이터가 구성되어 있을 경우

In [17]:
df2 = pd.read_csv("datas2/반별_점수_type2.csv", encoding = "euc-kr")
df2.head()

,A반,B반
0,73,63.0
1,69,56.0
2,71,73.0
3,71,61.0
4,73,55.0


In [18]:
# 길이가 달라서 결측이 발생할 수 있으므로, 결측을 제거한 뒤 각 컬럼을 group_A와 group_B에 저장
group_A = df2['A반'].dropna().values
group_B = df2['B반'].dropna().values

In [19]:
print(group_A)
print(group_B)

[73 69 71 71 73 67 73 69 62 74 68 66 70 82 70 65 76 73 58 81]
[63. 56. 73. 61. 55. 77. 75. 65. 61. 55.]
